bundle-examples/bundle-sequential/introduction/a6e1.olx

```diff
@@ -0,0 +1 @@
+<html id="a6e14e9cc2e74ec197609d6955be63c8" filename="/introduction/a6e1.html"/>
```

Comment:
So this is probably out of scope for this PR, but I sometimes think it would be nice if a unit could be some kind of non-XML container file format, like how YAML can embed multiple documents in a single file, but more general. Almost like MIME multipart messages, e.g.:

```
MIME-Version: 1.0
Content-Type: multipart/mixed; boundary=xblock
<insert other headers here for vertical info>

--xblock
Content-Type: text/html
XBlock-Type: html5

<html tags we don't have to worry about being valid XML>

--xblock
Content-Type: text/markdown
XBlock-Type: problem

(capa markdown text here)
```

Reply:
@bradenmacdonald Usually the convention is to use `url_name=<id>` to refer to nested blocks. I have replaced that with `path=<path to olx file>`. The `filename` attribute is `html` block specific (and perhaps to blocks which inherit from it) so haven't changed it; though the identifier will need to be the full path.

Adding to LX-54.

Reply:
Can we use XInclude for the file includes? Having Blockstore understand how to follow and piece together an HTML block based on various attributes sounds like it's getting into XBlock runtime territory, but XInclude is a standard spec that lxml supports.

Reply:
I [tried to parse the HTML content to append it to the OLX node](https://github.com/open-craft/openedx-blockstore-relay/pull/1/commits/b18ab614a286bdb841adff84b1556304414a8f4b#diff-e44130bcee7479b4f9a12331518df666R68), and found that many of the HTML blocks in the DemoX course at least don't parse as valid XML.

I would expect that XInclude-ing that content would have the same issue?

Reply:
@ormsbee Since the content files will be stored in separate storage and Blockstore will not be reading them (just saving and tracking URLs), I do not see where Blockstore will get the chance to piece together the OLX?

Having said so I see it makes sense to move to a standard attribute for the embedded content for the `html` block. Should we add using `xinclude` in such scenarios as a recommendation for other blocks to adopt as well?

@pomegranited Yes, that is not going to work most of the time. `xinclude` will not have the same issue since we can use [`parse=text`](https://www.w3.org/TR/xinclude/#include_element). For now I would recommend saving the html in a separate file as is done in course exports.

Reply:
I suppose it depends on how much we want Blockstore itself to be able to piece things together. When doing rendering for preview or for the student view, we're probably going to want the entire Unit most of the time. If stitching together the resources that are supposed to be shown inline (so things like HTML modules, not subtitles) requires XBlock runtime understanding, then either:

1. The XBlock runtime is coupled to Blockstore and runs in-proc -- which we explicitly want to avoid.
2. It's going to have to make multiple network roundtrips to get the pieces it needs to stitch together.
3. It's all taken care of in a compile phase that sits with an XBlock runtime somewhere else.

But I wonder if we can get make Blockstore just smart enough to pull these things inline without requiring an XBlock runtime. So we use XInclude with `parse=text` to pull in things like HTML and later Markdown and Python files in Capa problems. Then a client is able to ask "Give me this Unit file, but processed with the includes." Blockstore can understand and cache that, and basically serve up whole all the OLX content of a Unit in one resource without actually needing to understand OLX. And since pretty much everything is immutable, we can aggressively cache it at the Blockstore layer.

Reply:
Hmm I see. I do not see any technical limitations with doing this since XInclude is a standard with library support. And the current "inclusion" mechanism actually is a bit weird because for nested XBlocks, the block element is present both in the parent OLX and the OLX of the block. So if we can move to something more standard that sounds like a plus to me.

Reply:
I think that combining files as text using Xincludes or option 3 (prepare+optimize+combine files during a compilation phase) make sense.

Reply:
With the data transfer PR, I have the option to nest the XBlock OLX directly inside the `vertical` OLX, essentially performing the compilation prior to transfer.

Or I can keep them separate, and use `XInclude` to include them.

Which mechanism is preferred?

Note: this is a separate question from how the `.html` content should be included in the HTML OLX.

Reply:
@pomegranited and I discussed this over chat. My recommendation was that for now, until there has been a decision on the OLX changes, sticking with the conventions used in course export sounds like the straightforward path to take.

Reply:
@ormsbee 
>  I sometimes think it would be nice if a unit could be some kind of non-XML container file format, like how YAML can embed multiple documents in a single file, but more general. Almost like MIME multipart messages

If we want to to ditch OLX, then why not use the  the JSON data (as YAML) as exported by the Course Blocks API?  We're already relying on it for native apps, and so will need to ensure that the same data can be exposed for Blockstore bundles.

I'm having a whinge, because I'm currently trying to wrench the transcript assets out of the video OLX.  The xmodule [inexplicably removes the nice `/static` path prefixes](https://github.com/edx/edx-platform/blob/b85d6f031c1d6590a184a873fcadb3b2d42024bd/common/lib/xmodule/xmodule/video_module/transcripts_utils.py#L383-L384), making it impossible to scrape out these paths using the [`static_replace`](https://github.com/edx/edx-platform/blob/b85d6f031c1d6590a184a873fcadb3b2d42024bd/common/djangoapps/static_replace/__init__.py#L151) mechanism that works for other xblocks.  But if I could have used `student_view_data` instead of OLX, I'd already have everything I need in structured form, and wouldn't need to scrape anything:
```json
{
   "student_view_data":{
      "duration":null,
      "all_sources":[
         "https://s3.amazonaws.com/edx-course-videos/mit-6002x/6002-Tutorial-00010_100.mov"
      ],
      "transcripts":{
         "en":"http://localhost:8000/courses/course-v1:edX+DemoX+Demo_Course/xblock/block-v1:edX+DemoX+Demo_Course+type@video+block@636541acbae448d98ab484b028c9a7f6/handler_noauth/transcript/download?lang=en"
      },
      "only_on_web":false,
      "encoded_videos":{
         "fallback":{
            "url":"https://s3.amazonaws.com/edx-course-videos/mit-6002x/6002-Tutorial-00010_100.mov",
            "file_size":0
         },
         "youtube":{
            "url":"https://www.youtube.com/watch?v=o2pLltkrhGM",
            "file_size":0
         }
      }
   }
}
```